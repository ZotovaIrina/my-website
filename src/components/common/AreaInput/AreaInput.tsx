import * as PropTypes from 'prop-types';
import * as React from 'react';
import FormCell, { IFormCell } from '../FormCell/FormCell';

export interface IAreaInputProps extends IFormCell {
  /** Specifies is the textarea should automatically get focus. It also works after initializing the component using the imperative approach 'false -> true'. */
  autoFocus?: boolean;
  /** The value attribute of underlying textarea. */
  inputValue: string | null;
  /** A callback triggered when focus is removed. */
  onBlur?: (event: React.FocusEvent<HTMLTextAreaElement>) => void;
  /** A callback fired when the value prop changes. */
  onChange?: (newValue: string | null, error?: string | null) => void;
  /** Text that will appear in an empty input. */
  placeHolder?: string;
  /** Specifies the visible number of lines in underlying textarea. */
  rows?: number;
  /** Visually highlights the component as in error. */
  showErrors?: boolean;
  /** Switches the input to read only which prevents editing the contents. */
  readOnly?: boolean;
  /** This callback exposes the input reference / DOM node to parent components. */
  getInputRef?: (elem: HTMLTextAreaElement) => void;
}

interface IAreaInputState {
  value: string;
}

/**
 * The HTML textarea with a label and error messaging.
 */
export default class AreaInput extends React.Component<IAreaInputProps, IAreaInputState> {
  public static defaultProps: Partial<IAreaInputProps> = {
    inputValue: '',
    showErrors: true
  };

  public static propTypes = {
    autoFocus: PropTypes.bool,
    inputValue: PropTypes.string,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    placeHolder: PropTypes.string,
    rows: PropTypes.number,
    showErrors: PropTypes.bool
  };

  public state: IAreaInputState = {
    value: this.props.inputValue || ''
  };

  public textInput: HTMLTextAreaElement | null = null;

  public componentWillReceiveProps(nextProps: IAreaInputProps) {
    if (nextProps.autoFocus && nextProps.autoFocus !== this.props.autoFocus) {
      this.doAutoFocus(nextProps.autoFocus);
    }

    if (nextProps.inputValue !== this.state.value) {
      this.setState({ value: nextProps.inputValue || '' });
    }
  }

  public componentDidMount() {
    if (this.props.autoFocus) {
      this.doAutoFocus(this.props.autoFocus);
    }
  }

  public render() {
    return (
      <FormCell className={this.props.className} errorText={this.props.errorText} isRequired={this.props.isRequired} label={this.props.label}>
        <textarea
          className="faux-text-area"
          onBlur={this.props.onBlur}
          onChange={this.onChange}
          placeholder={this.props.placeHolder}
          readOnly={this.props.readOnly}
          ref={this.getInputRef}
          rows={this.props.rows}
          value={this.state.value}
        />
      </FormCell>
    );
  }

  private getInputRef = (input: HTMLTextAreaElement) => {
    this.textInput = input;
    if (this.props.getInputRef) {
      this.props.getInputRef(input);
    }
  };

  public onChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    let errorMessage: string | null = null;
    if (this.props.isRequired && !e.target.value && this.props.showErrors) {
      errorMessage = `${this.props.label} cannot be blank`;
    }
    this.setState({
      value: e.target.value
    });
    if (this.props.onChange) {
      this.props.onChange(e.target.value, errorMessage);
    }
  };

  public doAutoFocus = (autofocus: boolean) => {
    if (autofocus && this.textInput) {
      this.textInput.focus();
    }
  };
}
