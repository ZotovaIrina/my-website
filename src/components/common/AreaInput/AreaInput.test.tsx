import { mount } from 'enzyme';
import { toMatchImageSnapshot } from 'jest-image-snapshot';
import * as React from 'react';
import { create } from 'react-test-renderer';
import { attachHookableNode, removeHookableNode } from '../../test-utilities/handleHookableNode';
import AreaInput, { IAreaInputProps } from './AreaInput';

expect.extend({ toMatchImageSnapshot });

describe('AreaInput', () => {
  /**
   * Functional test suite utilitizes
   * enzyme.mount(..., {attachTo}) and attachHookableNode(...)
   * to ensure full DOM structure and React lifecycle avalability.
   * This will execute as mutch as possible from component
   * lifecycle which will force errors to surface.
   *
   * Not using enzyme.shallow by purpouse on components.
   * Would use it for views consisted of multiple compoments
   * to confirm that general structure is in place.
   * Checking if html structure of component is in place
   * should be handled during development and then sealed with snapshots.
   *
   * Also not testing for falsy scenarios like providing wrong
   * prop value. This is to be picked up on component level
   * with TypeScript and PropTypes.
   */
  describe('Functional', () => {
    const nodeName = 'hookableNode';
    let hookableNode: HTMLDivElement;

    beforeEach(() => {
      hookableNode = attachHookableNode(nodeName);
    });

    afterEach(() => {
      removeHookableNode(hookableNode);
    });

    it('uses props.autoFocus', () => {
      const props: IAreaInputProps = {
        autoFocus: true,
        inputValue: 'initialValue',
        onBlur: jest.fn()
      };

      const comp = mount(<AreaInput {...props} />, { attachTo: window[nodeName] });
      const textarea = comp.find('textarea').instance();
      expect(textarea).toBe(document.activeElement);
      comp.unmount();
    });

    it('calls props.onBlur', () => {
      const props: IAreaInputProps = {
        autoFocus: true,
        inputValue: 'initialValue',
        onBlur: jest.fn()
      };

      const comp = mount(<AreaInput {...props} />, { attachTo: window[nodeName] });
      const textarea = (comp.find('textarea').instance() as unknown) as HTMLTextAreaElement;
      textarea.blur();
      expect(props.onBlur).toBeCalledTimes(1);
      comp.unmount();
    });

    it('calls props.onChange', () => {
      const props: IAreaInputProps = {
        inputValue: 'initialValue',
        onChange: jest.fn()
      };

      const comp = mount(<AreaInput {...props} />);
      const textarea = comp.find('textarea');
      const instance = (textarea.instance() as unknown) as HTMLTextAreaElement;
      expect(instance.value).toBe('initialValue');
      textarea.simulate('change', { target: { value: 'newValue' } });
      expect(instance.value).toBe('newValue');
      expect(props.onChange).toBeCalledTimes(1);
      expect(props.onChange).toHaveBeenCalledWith('newValue', null);
      comp.unmount();
    });

    it('should call getInputRef when set', () => {
      const callback = jest.fn();
      mount(<AreaInput getInputRef={callback}/>);
      expect(callback).toHaveBeenCalledWith(expect.anything());
    });


  });

  /**
   * Snapshot tests is to confirm if component was rendered properly
   * and to "seal" it. If snapshot changes it forces dev to reevaluate output.
   *
   * Provide scenarios with:
   * - only required props
   * - all props
   * - mix of above when mixed scenarios are possible
   *
   * It uses create(...) from react-test-renderer
   * to provide almost clean HTML output for easy debug.
   */
  describe('Snapshot', () => {
    it('required props', () => {
      const props: IAreaInputProps = {
        inputValue: 'inputValue'
      };

      const tree = create(<AreaInput {...props} />);
      expect(tree).toMatchSnapshot();
    });

    it('all props', () => {
      const props: IAreaInputProps = {
        autoFocus: true,
        className: 'className',
        errorText: 'errorText',
        hintText: 'hintText',
        inputValue: 'inputValue',
        isRequired: true,
        label: 'label',
        onBlur: () => void 0,
        onChange: () => void 0,
        placeHolder: 'placeholder',
        rows: 1,
        showErrors: true,
        wrapInput: true
      };
      const tree = create(<AreaInput {...props} />);
      expect(tree).toMatchSnapshot();
    });
  });

  /**
   * UnitJS test suite tests only single pure javascript functions.
   */
  // describe('UnitJS', () => {
  //     it(('someFunctionName should return ... for ...') => {
  //         const opts: ISomeFunctionName = {};
  //         const expected = ...;
  //         const result = someFunctionName(opts);
  //         expect(result).toBe(expected);
  //     })
  // });
});
