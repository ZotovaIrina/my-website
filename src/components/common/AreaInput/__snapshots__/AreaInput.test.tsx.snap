// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreaInput Snapshot all props 1`] = `
<div
  className="form-cell form-cell--error className"
>
  <span
    className="form-cell__label"
  >
    label *
  </span>
  <textarea
    className="faux-text-area"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="placeholder"
    rows={1}
    value="inputValue"
  />
  <span
    className="form-cell__error"
  >
    errorText
  </span>
</div>
`;

exports[`AreaInput Snapshot required props 1`] = `
<div
  className="form-cell"
>
  <textarea
    className="faux-text-area"
    onChange={[Function]}
    value="inputValue"
  />
</div>
`;
